File: src/dotnet/QsoRipper.Engine.DotNet/QrzSecretStore.cs
using KeySharp;

namespace QsoRipper.Engine.DotNet;

internal enum QrzSecret
{
    XmlPassword,
    LogbookApiKey,
}

internal interface IQrzSecretStore
{
    string? Get(QrzSecret secret);

    void Set(QrzSecret secret, string value);
}

internal sealed class NullQrzSecretStore : IQrzSecretStore
{
    public string? Get(QrzSecret secret) => null;

    public void Set(QrzSecret secret, string value)
    {
    }
}

internal sealed class PlatformQrzSecretStore : IQrzSecretStore
{
    private const string Package = "com.treitforge.qsoripper";
    private const string Service = "qrz";

    public string? Get(QrzSecret secret)
    {
        try
        {
            return Keyring.GetPassword(Package, Service, GetAccount(secret));
        }
        catch (KeyringException exception) when (exception.Type == ErrorType.NotFound)
        {
            return null;
        }
        catch (Exception exception)
        {
            throw new QrzSecretStoreException("The platform credential store is unavailable.", exception);
        }
    }

    public void Set(QrzSecret secret, string value)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(value);

        try
        {
            Keyring.SetPassword(Package, Service, GetAccount(secret), value);
        }
        catch (Exception exception)
        {
            throw new QrzSecretStoreException("The platform credential store did not save the QRZ secret.", exception);
        }
    }

    private static string GetAccount(QrzSecret secret) =>
        secret switch
        {
            QrzSecret.XmlPassword => "xml-password",
            QrzSecret.LogbookApiKey => "logbook-api-key",
            _ => throw new ArgumentOutOfRangeException(nameof(secret)),
        };
}

internal sealed class QrzSecretStoreException : InvalidOperationException
{
    public QrzSecretStoreException()
    {
    }

    public QrzSecretStoreException(string message)
        : base(message)
    {
    }

    public QrzSecretStoreException(string message, Exception innerException)
        : base(message, innerException)
    {
    }
}

File: src/rust/qsoripper-server/src/qrz_secret_store.rs
#[cfg(not(test))]
const PACKAGE: &str = "com.treitforge.qsoripper";
#[cfg(not(test))]
const SERVICE: &str = "qrz";
#[cfg(all(not(test), target_os = "macos"))]
const FULL_SERVICE: &str = "com.treitforge.qsoripper.qrz";

#[derive(Clone, Copy, Debug)]
pub(crate) enum QrzSecret {
    XmlPassword,
    LogbookApiKey,
}

impl QrzSecret {
    fn account(self) -> &'static str {
        match self {
            Self::XmlPassword => "xml-password",
            Self::LogbookApiKey => "logbook-api-key",
        }
    }
}

pub(crate) trait QrzSecretStore: Send + Sync {
    fn get(&self, secret: QrzSecret) -> Result<Option<String>, String>;
    fn set(&self, secret: QrzSecret, value: &str) -> Result<(), String>;
}

#[cfg(not(test))]
pub(crate) struct PlatformQrzSecretStore;

#[cfg(not(test))]
impl PlatformQrzSecretStore {
    pub(crate) fn new() -> Self {
        Self
    }

    fn entry(secret: QrzSecret) -> Result<keyring::Entry, String> {
        let account = secret.account();

        #[cfg(target_os = "windows")]
        let entry = keyring::Entry::new(&format!("{SERVICE}/{account}"), PACKAGE);

        #[cfg(target_os = "macos")]
        let entry = keyring::Entry::new(FULL_SERVICE, account);

        #[cfg(all(unix, not(target_os = "macos")))]
        let entry = keyring::Entry::new(SERVICE, account);

        #[cfg(not(any(target_os = "windows", target_os = "macos", unix)))]
        return Err("The platform credential store is not supported.".to_string());

        entry.map_err(|error| format!("Platform credential store is unavailable: {error}"))
    }
}

#[cfg(not(test))]
impl QrzSecretStore for PlatformQrzSecretStore {
    fn get(&self, secret: QrzSecret) -> Result<Option<String>, String> {
        match Self::entry(secret)?.get_password() {
            Ok(value) => Ok(Some(value)),
            Err(keyring::Error::NoEntry) => Ok(None),
            Err(error) => Err(format!(
                "Failed to read the platform credential store: {error}"
            )),
        }
    }

    fn set(&self, secret: QrzSecret, value: &str) -> Result<(), String> {
        let entry = Self::entry(secret)?;
        entry
            .set_password(value)
            .map_err(|error| format!("Failed to write the platform credential store: {error}"))?;

        #[cfg(all(unix, not(target_os = "macos")))]
        entry
            .inner
            .update_attributes(&std::collections::HashMap::from([("xdg:schema", PACKAGE)]))
            .map_err(|error| {
                format!("Failed to set the platform credential store schema: {error}")
            })?;

        Ok(())
    }
}

#[cfg(test)]
#[derive(Default)]
pub(crate) struct MemoryQrzSecretStore {
    values: std::sync::RwLock<std::collections::HashMap<&'static str, String>>,
}

#[cfg(test)]
impl QrzSecretStore for MemoryQrzSecretStore {
    fn get(&self, secret: QrzSecret) -> Result<Option<String>, String> {
        let values = self
            .values
            .read()
            .map_err(|_| "Memory credential store read failed.".to_string())?;
        Ok(values.get(secret.account()).cloned())
    }

    fn set(&self, secret: QrzSecret, value: &str) -> Result<(), String> {
        let mut values = self
            .values
            .write()
            .map_err(|_| "Memory credential store write failed.".to_string())?;
        values.insert(secret.account(), value.to_string());
        Ok(())
    }
}
